server/src/main/java/org/opensearch/indices/IndicesRequestCache.java

```diff
@@ -228,19 +240,31 @@ BytesReference getOrCompute(
         final Key key = new Key(((IndexShard) cacheEntity.getCacheIdentity()).shardId(), cacheKey, readerCacheKeyId);
         Loader cacheLoader = new Loader(cacheEntity, loader);
         BytesReference value = cache.computeIfAbsent(key, cacheLoader);
-        if (cacheLoader.isLoaded()) {
-            cacheEntity.onMiss();
-            // see if it's the first time we see this reader, and make sure to register a cleanup key
-            CleanupKey cleanupKey = new CleanupKey(cacheEntity, readerCacheKeyId);
-            if (!registeredClosedListeners.containsKey(cleanupKey)) {
-                Boolean previous = registeredClosedListeners.putIfAbsent(cleanupKey, Boolean.TRUE);
-                if (previous == null) {
-                    OpenSearchDirectoryReader.addReaderCloseListener(reader, cleanupKey);
+        CleanupKey cleanupKey = null;
+        try {
+            if (cacheLoader.isLoaded()) {
+                cacheEntity.onMiss();
+                // see if it's the first time we see this reader, and make sure to register a cleanup key
+                cleanupKey = new CleanupKey(cacheEntity, readerCacheKeyId);
+                if (!registeredClosedListeners.containsKey(cleanupKey)) {
+                    Boolean previous = registeredClosedListeners.putIfAbsent(cleanupKey, Boolean.TRUE);
+                    if (previous == null) {
+                        OpenSearchDirectoryReader.addReaderCloseListener(reader, cleanupKey);
```

Comment:
If this is a bug indeed, I wonder how come we never any exceptions. I think we are missing something here.
I think in case an exception is thrown from this cache, OS just swallows it and goes with normal operation by running the query directly without caching it. Can we verify this behaviour?